src/H5EA.c

```diff
@@ -990,27 +1015,26 @@ BEGIN_FUNC(PRIV, ERR, int, H5_ITER_CONT, H5_ITER_ERROR,
 
     /* Allocate space for a native array element */
     if (NULL == (elmt = H5FL_BLK_MALLOC(ea_native_elmt, ea->hdr->cparam.cls->nat_elmt_size)))
-        H5E_THROW(H5E_CANTALLOC, "memory allocation failed for extensible array element")
+        HGOTO_ERROR(H5E_EARRAY, H5E_CANTALLOC, H5_ITER_ERROR,
+                    "memory allocation failed for extensible array element")
 
     /* Iterate over all elements in array */
-    for (u = 0; u < ea->hdr->stats.stored.max_idx_set && cb_ret == H5_ITER_CONT; u++) {
+    for (u = 0; u < ea->hdr->stats.stored.max_idx_set && ret_value == H5_ITER_CONT; u++) {
         /* Get array element */
         if (H5EA_get(ea, u, elmt) < 0)
-            H5E_THROW(H5E_CANTGET, "unable to delete fixed array")
+            HGOTO_ERROR(H5E_EARRAY, H5E_CANTGET, H5_ITER_ERROR, "unable to delete fixed array")
 
         /* Make callback */
-        if ((cb_ret = (*op)(u, elmt, udata)) < 0) {
-            H5E_PRINTF(H5E_BADITER, "iterator function failed");
-            H5_LEAVE(cb_ret)
-        } /* end if */
-    }     /* end for */
-
-    CATCH
+        if ((ret_value = (*op)(u, elmt, udata)) < 0)
+            HGOTO_DONE(ret_value)
```

Comment:
Should this be an HGOTO_ERROR with the old "iterator function failed" message rather than a goto_done?

Reply:
This should be an HERROR with a break out of the loop or HGOTO_DONE, similar to other iterators in the library.   For a good example, see the H5A__iterate_common / H5A__iterate routines in src/H5Aint.c

Reply:
Fixed